sdk.go

```diff
@@ -0,0 +1,188 @@
+package tflog
+
+import (
+	"context"
+
+	"github.com/hashicorp/go-hclog"
+)
+
+const newSDKSubsystemLoggerWarning = "This log was generated by an SDK subsystem logger that wasn't created before being used. Use tflog.NewSDKSubsystem to create this logger before it is used."
+
+func getSDKRootLogger(ctx context.Context) hclog.Logger {
+	logger := ctx.Value(sdkRootLoggerKey)
+	if logger == nil {
+		return nil
+	}
+	return logger.(hclog.Logger)
+}
+
+func setSDKRootLogger(ctx context.Context, logger hclog.Logger) context.Context {
+	return context.WithValue(ctx, sdkRootLoggerKey, logger)
+}
+
+func getSDKSubsystemLogger(ctx context.Context, subsystem string) hclog.Logger {
+	logger := ctx.Value(sdkRootLoggerKey + loggerKey("."+subsystem))
+	if logger == nil {
+		return nil
+	}
+	return logger.(hclog.Logger)
+}
+
+func setSDKSubsystemLogger(ctx context.Context, subsystem string, logger hclog.Logger) context.Context {
+	return context.WithValue(ctx, sdkRootLoggerKey+loggerKey("."+subsystem), logger)
+}
+
+func NewSDK(ctx context.Context, options ...Option) context.Context {
```

Comment:
Should the SDK functionality be broken into a separate package to prevent clogging up the root namespace? https://pkg.go.dev/github.com/hashicorp/terraform-plugin-log@v0.0.0-20210430130036-e68e6ce2a2c8

That package can then be documented as "only expected to be implemented by the Terraform Plugin SDK" and help simplify what provider developers are looking at when consuming the package. 😄

Reply:
Probably! I like that solution a lot.